api/query_engine.go

```diff
@@ -15,84 +17,111 @@ var (
 	reqRenderPointsFetched = stats.NewMeter32("api.request.render.points_fetched", false)
 	// metric api.request.render.series is the number of points the request will return.
 	reqRenderPointsReturned = stats.NewMeter32("api.request.render.points_returned", false)
+
+	errUnSatisfiable = errors.New("request cannot be satisfied due to lack of available retentions")
 )
 
 // alignRequests updates the requests with all details for fetching, making sure all metrics are in the same, optimal interval
-// luckily, all metrics still use the same aggSettings, making this a bit simpler
 // note: it is assumed that all requests have the same from & to.
 // also takes a "now" value which we compare the TTL against
-func alignRequests(now uint32, reqs []models.Req, s mdata.AggSettings) ([]models.Req, error) {
+func alignRequests(now uint32, reqs []models.Req) ([]models.Req, error) {
 
 	tsRange := (reqs[0].To - reqs[0].From)
 
-	// prefer raw first and foremost.
-	// if raw interval doesn't retain data long enough, we must
-	// find the highest res rollup archive that retains all the data we need.
+	var listIntervals []uint32
+	var seenIntervals = make(map[uint32]struct{})
+
+	// set preliminary settings. may be adjusted further down
+	// but for now:
+	// for each req, find the highest res archive
+	// (starting with raw, then rollups in decreasing precision)
+	// that retains all the data we need.
 	// fallback to lowest res option (which *should* have the longest TTL)
 
-	archive := 0
-	ttl := s.RawTTL
+	for i := range reqs {
+		req := &reqs[i]
+		retentions := mdata.GetRetentions(req.SchemaI)
+		req.Archive = -1
 
-	if now-s.RawTTL > reqs[0].From {
-		for i, agg := range s.Aggs {
+		for i, ret := range retentions {
 			// skip non-ready option.
-			if !agg.Ready {
+			if !ret.Ready {
 				continue
 			}
-			archive = i + 1
-			ttl = agg.TTL
-			if now-agg.TTL <= reqs[0].From {
+
+			req.Archive = i
+			req.TTL = uint32(ret.MaxRetention())
+			req.ArchInterval = uint32(ret.SecondsPerPoint())
+
+			if now-req.TTL <= req.From {
 				break
 			}
 		}
 
-	}
+		if req.Archive == -1 {
+			return nil, errUnSatisfiable
```

Comment:
IMHO yes. this is such a rare case (basically all archives have ready=false). it's okay to make this problem very clear so that the admin can fix it. once we have https://github.com/grafana/grafana/issues/6448 we could make this a bit nicer (it would allow us to show a warning/error and still return the data we can)